CUDA support in the CSR layout:  CUDA addmm

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57403 CUDA support in the CSR layout:  CUDA addmm**
* #57275 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* #57274 CUDA support in the CSR layout: constructors

This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

aocsa committed:
CUDA support in the CSR layout: matvec

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

aocsa committed:
addressed review comments on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
review comments addressed on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
minor lint fix on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
addressed review comments - 2 on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
addmm fix on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
addmm cuda on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
rebased/addressed review comments - 3 on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
rebased/last feedback addressed on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
run ci test on "CUDA support in the CSR layout: matvec"


Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
run CI/last feedback addressed on "CUDA support in the CSR layout: matvec"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
rebased needed on "CUDA support in the CSR layout: matvec"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
rebased needed : on "CUDA support in the CSR layout: matvec"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
support for MKL matvec on "CUDA support in the CSR layout: matvec"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
clean stack on "CUDA support in the CSR layout: matvec"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa commented:
I rebased this PR and polished a few things. As a gentle reminder, this PR is still waiting for a final review before merging, and I won't be able to work on this after this month. cc @pearu, @ezyang

pearu commented on test/test_sparse_csr.py:
```diff
@@ -396,6 +394,34 @@ def test_csr_matvec(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, "mv: expected"):
                 csr.matmul(bad_vec)
 
+    @onlyCUDA
```
Test this for CPU as well?

aocsa replied:
Yes, This PR was waiting the merge of PR https://github.com/pytorch/pytorch/pull/58768

pearu commented on test/test_sparse_csr.py:
```diff
@@ -396,6 +394,34 @@ def test_csr_matvec(self, device, dtype):
+    @dtypes(torch.double)
+    def test_mm(self, device, dtype):
+        def test_shape(di, dj, dk, nnz):
+            x = self.genSparseCSRTensor((di, dj), nnz, device=device, dtype=dtype, index_dtype=torch.int32)
```
For completeness, CSR tensors should be tested for both int32 and int64 indices.

pearu commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,76 @@ namespace {
+        } else if (!is_same_tensor(t, r_)) {
+          r_.copy_(t);
+        }
+
```
Start `if (nnz > 0)` block from here.

aocsa committed:
rebased needed after csr addmm fix is merged on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,76 @@ namespace {
+            cast_beta,
+            r__.data_ptr<scalar_t>(),
+            r__.stride(1));
+        }
```
The implementation of addmm seems to be incomplete: if `nnz==0`, currently, the result will be `0`, but it should be `beta * r_`.

pearu reviewed (changes requested):
Thanks, @aocsa !

The implementation of addmm seems to be incomplete, IIUC: when `nnz == 0`, the result is all zeros but it ought to be `beta * input`.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -33,6 +35,82 @@ using namespace at::sparse_csr;
+
+  if (sparse.crow_indices().scalar_type() != kInt) {
+    TORCH_WARN(
+        "Pytorch is compiled with MKL LP64 and will convert crow_indices to int32.");
+  }
+  if (sparse.col_indices().scalar_type() != kInt) {
+    TORCH_WARN(
+        "Pytorch is compiled with MKL LP64 and will convert col_indices to int32.");
```
Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices. 

https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L200

I will remove these warnings to be consistent with .
https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L62

pearu replied:
> Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices.

See https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L125-L126

IIUC, the task boils down to generalizing `template<typename T> void _csrmm2...` to `template<typename T, typename I>
void _csrmm2` where `I` is index type.

If this makes sense, add `TODO` comment for follow-up work.

aocsa replied:
Sure, make sense.

aocsa replied:
I will create an issue for this too as it could affect other things.  
NVIDIA docs mention that CUSPARSE_INDEX_64I can be used but for CUSPARSE_SPMM_CSR_ALG2/CUSPARSE_SPMM_CSR_ALG3 currently is it uses CUSPARSE_CSRMM_ALG1
https://docs.nvidia.com/cuda/cusparse/index.html

> cusparseSpMM supports the following index type for representing the sparse matrix matA:
> 32-bit indices (CUSPARSE_INDEX_32I)
> 64-bit indices (CUSPARSE_INDEX_64I) only with CUSPARSE_SPMM_COO_ALG4, CUSPARSE_SPMM_CSR_ALG2, and CUSPARSE_SPMM_CSR_ALG3 algorithms

aocsa replied:
Mentioned here: https://github.com/pytorch/pytorch/issues/58899

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,76 @@ namespace {
+          if(dense.stride(0) == 1 && dense.stride(1) == dense.size(0)) {
+            transpose_dense = 'n';
+            dense_ = dense;
+          } else if(dense.stride(1) == 1 && dense.stride(0) != dense.size(1)) {
```
Seems ok. Moreover I added a test for transposed input tensors

pearu replied:
How does this seem ok? For instance, it does not match the same pattern used above: `if(dense.stride(0) == 1 && dense.stride(1) == dense.size(0))`..

So, let's analyze this: `dense` is a 2D tensor and if its last stride (`stride(1)`) is 1, then the first stride must be `size(1) * s` where `s` is some positive integer. Here, `dense` must be contiguous, hence, `s == 1` and `stride(0) == size(1)`, in general. As a result, the code block under `if(dense.stride(1) == 1 && dense.stride(0) != dense.size(1))` is never reached.

The test cases most likely use the `else` code block, so tests will not discover this bug.

aocsa replied:
Sorry I misunderstood the comment. Now 2nd condition is `else if(dense.stride(1) == 1 && dense.stride(0) == dense.size(1)) `

aocsa committed:
fix for case nnz==0 in addmm on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa commented:
> Thanks, @aocsa !
> 
> The implementation of addmm seems to be incomplete, IIUC: when `nnz == 0`, the result is all zeros but it ought to be `beta * input`.

Thank u @pearu.  I addressed the feedback comments and added new test cases.  Let me know your thoughts.

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,75 @@ namespace {
+        } else if (!is_same_tensor(t, r_)) {
+          r_.copy_(t);
+        }
+        if (nnz > 0) {
```
With `if (nnz == 0)` used in the caller functions, this `if` condition can now be eliminated.
Use `TORCH_INTERNAL_ASSERT(nnz > 0)` instead.

pearu commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,75 @@ namespace {
+          if(r_.stride(0) == 1 && r_.stride(1) == r_.size(0)) {
+            r__ = r_;
+          } else {
+            // TODO: how... strange
```
Not sure what this TODO is about. What the code does below, is just ensuring that `r__` is Fortran contiguous.
Perhaps add a comment explaining why this is preferred?

pearu reviewed (approved):
I have a couple of nit comments. Otherwise, LGTM. Thanks, @aocsa!

aocsa committed:
last updates to csr cuda addmm on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa commented:
> I have a couple of nit comments. Otherwise, LGTM. Thanks, @aocsa!

Thanks @pearu.

aocsa commented:
I think this stack  PR needs to re-import PR to be merged

aocsa committed:
minor update to PR 57275 on "CUDA support in the CSR layout:  CUDA addmm"


This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

[ghstack-poisoned]

aocsa commented:
Closing this PR too :(. Stack PR was poisoned after first  PR in the stack was reverted. New PR is here:
https://github.com/pytorch/pytorch/pull/59004

bhosmer commented:
new new PR: new new PR: #59012 (thanks for restacking @aocsa)
